404 pages don't work when a site has Functions but no SSR/DSG pages

Deploying a Gatsby 4 site with functions defined in the `api` folder breaks 404 handling in production.

If trying to access a non-existent page like `/404-test` triggers the below error: 
```
{"errorType":"Runtime.ImportModuleError","errorMessage":"Error: Cannot find module '/var/task/.cache/page-ssr'\nRequire stack:\n- /var/task/.netlify/functions-internal/__dsg/__dsg.js\n- /var/task/__dsg.js\n- /var/runtime/UserFunction.js\n- /var/runtime/index.js","trace":["Runtime.ImportModuleError: Error: Cannot find module '/var/task/.cache/page-ssr'","Require stack:","- /var/task/.netlify/functions-internal/__dsg/__dsg.js","- /var/task/__dsg.js","- /var/runtime/UserFunction.js","- /var/runtime/index.js","    at _loadUserApp (/var/runtime/UserFunction.js:202:13)","    at Object.module.exports.load (/var/runtime/UserFunction.js:242:17)","    at Object.<anonymous> (/var/runtime/index.js:43:30)","    at Module._compile (internal/modules/cjs/loader.js:1085:14)","    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)","    at Module.load (internal/modules/cjs/loader.js:950:32)","    at Function.Module._load (internal/modules/cjs/loader.js:790:12)","    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)","    at internal/main/run_main_module.js:17:47"]}
```

If you rename the gatsby `api` folder to `_api` and redeploy 404 pages start working correctly.

gatsby-plugin-netlify version 4.1 is installed and enabled.

--- a/README.md
+++ b/README.md
@@ -45,7 +45,7 @@ manually, you have two options:
 
 You should also install the Gatsby plugin
 [gatsby-plugin-netlify](https://www.gatsbyjs.org/plugins/gatsby-plugin-netlify/).
-This is required for SSR pages, and adds support for Gatsby redirects and asset
+This is required for SSR and DSG pages, and adds support for Gatsby redirects and asset
 caching rules:
 
 1. Add the package as a dependency:

--- a/plugin/package.json
+++ b/plugin/package.json
@@ -54,6 +54,7 @@
     "node-stream-zip": "^1.15.0",
     "pathe": "^0.2.0",
     "pretty-bytes": "^5.6.0",
+    "semver": "^7.3.5",
     "tempy": "^1.0.0"
   },
   "devDependencies": {
@@ -61,6 +62,7 @@
     "@netlify/build": "^26.5.1",
     "@types/fs-extra": "^9.0.12",
     "@types/multer": "^1.4.7",
+    "@types/semver": "^7.3.9",
     "gatsby": "^4.9.0",
     "npm-run-all": "^4.1.5",
     "rimraf": "^3.0.2",

--- a/plugin/src/helpers/config.ts
+++ b/plugin/src/helpers/config.ts
@@ -6,6 +6,8 @@ import { stripIndent } from 'common-tags'
 import fs, { existsSync } from 'fs-extra'
 import type { GatsbyConfig, PluginRef } from 'gatsby'
 
+import { checkPackageVersion } from './files'
+
 export async function spliceConfig({
   startMarker,
   endMarker,
@@ -38,11 +40,9 @@ export async function spliceConfig({
   return fs.writeFile(fileName, out)
 }
 
-function loadGatsbyConfig({ utils, publish }): GatsbyConfig | never {
-  const gatsbyConfigFile = path.resolve(
-    getGatsbyRoot(publish),
-    'gatsby-config.js',
-  )
+function loadGatsbyConfig({ gatsbyRoot, utils }): GatsbyConfig | never {
+  const gatsbyConfigFile = path.resolve(gatsbyRoot, 'gatsby-config.js')
+
   if (!existsSync(gatsbyConfigFile)) {
     return {}
   }
@@ -65,14 +65,25 @@ function hasPlugin(plugins: PluginRef[], pluginName: string): boolean {
   )
 }
 
-export function checkGatsbyConfig({ utils, netlifyConfig }): void {
+export async function checkConfig({ utils, netlifyConfig }): Promise<void> {
+  const gatsbyRoot = getGatsbyRoot(netlifyConfig.build.publish)
+
   // warn if gatsby-plugin-netlify is missing
   const gatsbyConfig = loadGatsbyConfig({
     utils,
-    publish: netlifyConfig.build.publish,
+    gatsbyRoot,
   })
 
-  if (!hasPlugin(gatsbyConfig.plugins, 'gatsby-plugin-netlify')) {
+  if (hasPlugin(gatsbyConfig.plugins, 'gatsby-plugin-netlify')) {
+    if (
+      // prettier-ignore
+      !(await checkPackageVersion(gatsbyRoot, 'gatsby-plugin-netlify', '>=4.2.0',))
+    ) {
+      console.error(
+        'The plugin `gatsby-plugin-netlify` does not support DSG, please update to >=4.2.0',
+      )
+    }
+  } else {
     console.error(
       'Please install `gatsby-plugin-netlify` and enable it in your gatsby-config.js. https://www.gatsbyjs.com/plugins/gatsby-plugin-netlify/',
     )

--- a/plugin/src/helpers/config.ts
+++ b/plugin/src/helpers/config.ts
@@ -1,3 +1,4 @@
+/* eslint-disable max-lines */
 import { EOL } from 'os'
 import path from 'path'
 import process from 'process'
@@ -76,8 +77,11 @@ export async function checkConfig({ utils, netlifyConfig }): Promise<void> {
 
   if (hasPlugin(gatsbyConfig.plugins, 'gatsby-plugin-netlify')) {
     if (
-      // prettier-ignore
-      !(await checkPackageVersion(gatsbyRoot, 'gatsby-plugin-netlify', '>=4.2.0',))
+      !(await checkPackageVersion(
+        gatsbyRoot,
+        'gatsby-plugin-netlify',
+        '>=4.2.0',
+      ))
     ) {
       console.error(
         'The plugin `gatsby-plugin-netlify` does not support DSG, please update to >=4.2.0',
@@ -173,3 +177,4 @@ export function shouldSkipFunctions(cacheDir: string): boolean {
 export function getGatsbyRoot(publish: string): string {
   return path.resolve(path.dirname(publish))
 }
+/* eslint-enable max-lines */

--- a/plugin/src/helpers/files.ts
+++ b/plugin/src/helpers/files.ts
@@ -1,8 +1,16 @@
 import os from 'os'
 import process from 'process'
 
-import { copyFile, ensureDir, existsSync, readFile, writeFile } from 'fs-extra'
+import {
+  copyFile,
+  ensureDir,
+  existsSync,
+  readFile,
+  writeFile,
+  readJson,
+} from 'fs-extra'
 import { dirname, join, resolve } from 'pathe'
+import semver from 'semver'
 
 const DEFAULT_LAMBDA_PLATFORM = 'linux'
 const DEFAULT_LAMBDA_ABI = '83'
@@ -68,6 +76,26 @@ export const findModuleFromBase = ({ paths, candidates }): string | null => {
     }
   }
   return null
+}
+
+export async function checkPackageVersion(
+  root: string,
+  name: string,
+  version: string,
+): Promise<boolean> {
+  const packagePath = findModuleFromBase({
+    paths: [root],
+    candidates: [name],
+  })
+
+  let packageObj: { version: string }
+  try {
+    packageObj = await readJson(`${packagePath}/package.json`)
+  } catch {
+    return false
+  }
+
+  return semver.satisfies(packageObj.version, version)
 }
 
 /**

--- a/plugin/src/helpers/files.ts
+++ b/plugin/src/helpers/files.ts
@@ -83,19 +83,15 @@ export async function checkPackageVersion(
   name: string,
   version: string,
 ): Promise<boolean> {
-  const packagePath = findModuleFromBase({
-    paths: [root],
-    candidates: [name],
-  })
-
-  let packageObj: { version: string }
   try {
-    packageObj = await readJson(`${packagePath}/package.json`)
+    const packagePath = require.resolve(`${name}/package.json`, {
+      paths: [root],
+    })
+    const packageObj = await readJson(packagePath)
+    return semver.satisfies(packageObj.version, version)
   } catch {
     return false
   }
-
-  return semver.satisfies(packageObj.version, version)
 }
 
 /**

--- a/plugin/src/index.ts
+++ b/plugin/src/index.ts
@@ -7,7 +7,7 @@ import { existsSync } from 'fs-extra'
 
 import { normalizedCacheDir, restoreCache, saveCache } from './helpers/cache'
 import {
-  checkGatsbyConfig,
+  checkConfig,
   mutateConfig,
   shouldSkipFunctions,
   spliceConfig,
@@ -32,7 +32,7 @@ export async function onPreBuild({
   }
   await restoreCache({ utils, publish: PUBLISH_DIR })
 
-  checkGatsbyConfig({ utils, netlifyConfig })
+  await checkConfig({ utils, netlifyConfig })
 }
 
 export async function onBuild({

--- a/plugin/src/index.ts
+++ b/plugin/src/index.ts
@@ -79,12 +79,6 @@ The plugin no longer uses this and it should be deleted to avoid conflicts.\n`)
     contents: '/api/* /.netlify/functions/__api 200',
     fileName: join(netlifyConfig.build.publish, '_redirects'),
   })
-
-  netlifyConfig.redirects.push({
-    from: '/*',
-    to: '/.netlify/builders/__dsg',
-    status: 200,
-  })
 }
 
 export async function onPostBuild({

